### PR TITLE
New extended resources kublet flag

### PR DESCRIPTION
**Pull request by sethp-nr**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Enables kubeadm configured nodes to start up advertising a specific set
of extended resources, instead of needing to wait for the node to
register in the API and send a specific PATCH to the node's status.

**Which issue(s) this PR fixes**:

None that I could find, but both myself and one other person have raised the question in the sig-node slack channel.

**Special notes for your reviewer**:

I ignored the very good advice to not implement new kubelet flags because it's not clear to me how to define which subset of nodes get which KubeletConfiguration(s).

That said, I'm happy to learn more about component configs if you would help guide me in moving this config there and attaching it to the right nodes via `kubeadm`.

**Does this PR introduce a user-facing change?**:

```release-note
Adds a --node-extended-resources flag to instruct the kubelet to register the node with statically defined extended resoures
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by dashpole** [comment]
Can you say a bit more about your use-case?  Why doesn't sending a PATCH work?

**Comment by sethp-nr** [comment]
Sure! We're using cluster-api to manage our clusters, which means that there's not necessarily a human operator standing by to issue the PATCH when a new node joins the cluster. And in the cases where there is a human being, we often fail to issue the right command 😄

What cluster-api does give us is a way to plumb extra kubelet args through the `kubeadm` config applied to those machines to ensure they come into the cluster with the right set of resources.

The alternative we considered was a NoSchedule taint + controller that watches for some particular node label, and issues the PATCH and removes the NoSchedule taint. But since the `kubelet` already has flags for taints and the node label, it made sense to us (and would be operationally simpler) for it to also understood its own extended resources.

**Comment by dashpole** [comment]
If a pod requests an extended resource, but the node object doesn't have it, will it still be scheduled?  I'm trying to figure out why the taint is required

File: cmd/kubelet/app/options/options.go [hunk 1]
```diff
@@ -243,6 +247,13 @@ func ValidateKubeletFlags(f *KubeletFlags) error {
 		return fmt.Errorf("unknown 'kubernetes.io' or 'k8s.io' labels specified with --node-labels: %v\n--node-labels in the 'kubernetes.io' namespace must begin with an allowed prefix (%s) or be in the specifically allowed set (%s)", unknownLabels.List(), strings.Join(kubeletapis.KubeletLabelNamespaces(), ", "), strings.Join(kubeletapis.KubeletLabels(), ", "))
 	}
 
+	for k, v := range f.NodeExtendedResources {
+		// This is possibly a little more restrictive then necessary, but all pure integers are valid Quantities
```
Comment (neolit123) [inline cmd/kubelet/app/options/options.go]: then -> than

File: cmd/kubelet/app/options/options.go [hunk 2]
```diff
@@ -243,6 +247,13 @@ func ValidateKubeletFlags(f *KubeletFlags) error {
...
+		if i, err := strconv.ParseInt(v, 10, 64); err != nil || i < 0 {
+			return fmt.Errorf("node extended resources must be whole 64-bit signed (but non-negative) integers. Instead we saw %s=%s", k, v)
```
Comment (neolit123) [inline cmd/kubelet/app/options/options.go]: excluding the personallization "we".
"whole" numbers implies non-negative integers, possibly adjust.

proposing:

> node extended resources must be 64bit signed (but positive) integers; saw %s=%s

**Comment by neolit123** [comment]
@sethp-nr thanks for the PR, was there a discussion about this in the cluster-api or kubernetes/kubernetes issue tracker already?

> What cluster-api does give us is a way to plumb extra kubelet args through the kubeadm config applied to those machines to ensure they come into the cluster with the right set of resources.

alternatively, if are willing to use https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm for v1alpha2 of cluster-api.

i think it will allow you to apply a PATCH as a post `kubeadm init/join` step using cloud-init scripts.

**Comment by mattjmcnaughton** [comment]
@sethp-nr welcome to the k8s community :)

Could I trouble you to link the slack conversation? This seems like a non-trivial feature to add, so I just want to make sure its been given the proper discussion before we get too far into the code.

cc @derekwaynecarr @dchen1107 would we want a KEP for this feature? I'm not sure so wanted to escalate to y'all :)

**Pull request closed without merging**